File: filecrypt.cs
// filecrypt.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;

class FileCrypt
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    const int SALT_LEN = 16;
    const int NONCE_LEN = 12;
    const int TAG_LEN = 16;
    const int KEY_LEN = 32;
    const int ITERATIONS = 100000;

    static byte[] DeriveKey(string password, byte[] salt)
    {
        using var rfc = new Rfc2898DeriveBytes(password, salt, ITERATIONS, HashAlgorithmName.SHA256);
        return rfc.GetBytes(KEY_LEN);
    }

    static void EncryptFile(string inputPath, string outputPath, string password)
    {
        byte[] salt = RandomNumberGenerator.GetBytes(SALT_LEN);
        byte[] nonce = RandomNumberGenerator.GetBytes(NONCE_LEN);
        byte[] key = DeriveKey(password, salt);

        byte[] plaintext = File.ReadAllBytes(inputPath);
        using var aes = new AesGcm(key);
        byte[] ciphertext = new byte[plaintext.Length];
        byte[] tag = new byte[TAG_LEN];
        aes.Encrypt(nonce, plaintext, ciphertext, tag);

        using var fs = new FileStream(outputPath, FileMode.Create);
        fs.Write(salt, 0, salt.Length);
        fs.Write(nonce, 0, nonce.Length);
        fs.Write(ciphertext, 0, ciphertext.Length);
        fs.Write(tag, 0, tag.Length);
        Console.WriteLine(Colorize($"File encrypted successfully: {outputPath}", "green"));
    }

    static void DecryptFile(string inputPath, string outputPath, string password)
    {
        byte[] data = File.ReadAllBytes(inputPath);
        if (data.Length < SALT_LEN + NONCE_LEN + TAG_LEN)
            throw new Exception("Invalid file format");

        byte[] salt = new byte[SALT_LEN];
        byte[] nonce = new byte[NONCE_LEN];
        Buffer.BlockCopy(data, 0, salt, 0, SALT_LEN);
        Buffer.BlockCopy(data, SALT_LEN, nonce, 0, NONCE_LEN);
        int cipherLen = data.Length - SALT_LEN - NONCE_LEN - TAG_LEN;
        byte[] ciphertext = new byte[cipherLen];
        byte[] tag = new byte[TAG_LEN];
        Buffer.BlockCopy(data, SALT_LEN + NONCE_LEN, ciphertext, 0, cipherLen);
        Buffer.BlockCopy(data, SALT_LEN + NONCE_LEN + cipherLen, tag, 0, TAG_LEN);

        byte[] key = DeriveKey(password, salt);
        using var aes = new AesGcm(key);
        byte[] plaintext = new byte[cipherLen];
        aes.Decrypt(nonce, ciphertext, tag, plaintext);

        File.WriteAllBytes(outputPath, plaintext);
        Console.WriteLine(Colorize($"File decrypted successfully: {outputPath}", "green"));
    }

    static void Main(string[] args)
    {
        if (args.Length < 4)
        {
            Console.WriteLine(Colorize("Usage: filecrypt encrypt|decrypt <input> <output> <password>", "yellow"));
            return;
        }
        string mode = args[0];
        string input = args[1];
        string output = args[2];
        string password = args[3];

        try
        {
            if (mode == "encrypt") EncryptFile(input, output, password);
            else if (mode == "decrypt") DecryptFile(input, output, password);
            else throw new Exception("Invalid mode. Use encrypt or decrypt.");
        }
        catch (Exception e)
        {
            Console.WriteLine(Colorize($"Error: {e.Message}", "red"));
        }
    }
}

File: filecrypt.java
// filecrypt.java
import javax.crypto.*;
import javax.crypto.spec.*;
import java.io.*;
import java.security.*;
import java.security.spec.*;

public class filecrypt {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String RED = "\u001B[91m";
    private static final String YELLOW = "\u001B[93m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static final int SALT_LEN = 16;
    private static final int NONCE_LEN = 12;
    private static final int TAG_LEN = 16;
    private static final int KEY_LEN = 32;
    private static final int ITERATIONS = 100000;

    private static byte[] deriveKey(String password, byte[] salt) throws Exception {
        SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
        KeySpec spec = new PBEKeySpec(password.toCharArray(), salt, ITERATIONS, KEY_LEN * 8);
        SecretKey tmp = factory.generateSecret(spec);
        return tmp.getEncoded();
    }

    private static void encryptFile(String inputPath, String outputPath, String password) throws Exception {
        SecureRandom rng = SecureRandom.getInstanceStrong();
        byte[] salt = new byte[SALT_LEN];
        byte[] nonce = new byte[NONCE_LEN];
        rng.nextBytes(salt);
        rng.nextBytes(nonce);

        byte[] key = deriveKey(password, salt);
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        GCMParameterSpec gcmSpec = new GCMParameterSpec(TAG_LEN * 8, nonce);
        cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key, "AES"), gcmSpec);

        byte[] plaintext = Files.readAllBytes(Paths.get(inputPath));
        byte[] ciphertext = cipher.doFinal(plaintext);

        try (FileOutputStream fos = new FileOutputStream(outputPath)) {
            fos.write(salt);
            fos.write(nonce);
            fos.write(ciphertext);
        }
        System.out.println(colorize("File encrypted successfully: " + outputPath, GREEN));
    }

    private static void decryptFile(String inputPath, String outputPath, String password) throws Exception {
        byte[] data = Files.readAllBytes(Paths.get(inputPath));
        if (data.length < SALT_LEN + NONCE_LEN + TAG_LEN)
            throw new Exception("Invalid file format");

        byte[] salt = new byte[SALT_LEN];
        byte[] nonce = new byte[NONCE_LEN];
        System.arraycopy(data, 0, salt, 0, SALT_LEN);
        System.arraycopy(data, SALT_LEN, nonce, 0, NONCE_LEN);
        int cipherLen = data.length - SALT_LEN - NONCE_LEN;
        byte[] ciphertext = new byte[cipherLen];
        System.arraycopy(data, SALT_LEN + NONCE_LEN, ciphertext, 0, cipherLen);

        byte[] key = deriveKey(password, salt);
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        GCMParameterSpec gcmSpec = new GCMParameterSpec(TAG_LEN * 8, nonce);
        cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key, "AES"), gcmSpec);

        byte[] plaintext = cipher.doFinal(ciphertext);
        Files.write(Paths.get(outputPath), plaintext);
        System.out.println(colorize("File decrypted successfully: " + outputPath, GREEN));
    }

    public static void main(String[] args) {
        if (args.length < 4) {
            System.out.println(colorize("Usage: java filecrypt encrypt|decrypt <input> <output> <password>", YELLOW));
            return;
        }
        String mode = args[0];
        String input = args[1];
        String output = args[2];
        String password = args[3];

        try {
            if (mode.equals("encrypt")) encryptFile(input, output, password);
            else if (mode.equals("decrypt")) decryptFile(input, output, password);
            else throw new Exception("Invalid mode. Use encrypt or decrypt.");
        } catch (Exception e) {
            System.err.println(colorize("Error: " + e.getMessage(), RED));
        }
    }
}
